### PR TITLE
Move hard and soft in ExternalDocumentsIds into single map

### PR DESCRIPTION
**Pull request by msvaljek**
# Move hard and soft in ExternalDocumentsIds into single map docids

## Related issue
Fixes [#76](https://github.com/meilisearch/milli/issues/76)

## What does this PR do?
- as described in the issue comments removes hard and soft and replaces with a single map called docids

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by msvaljek** [comment]
I'm not sure how to do the benchmark.

I tried https://github.com/meilisearch/milli/tree/main/benchmarks#on-your-machine

The first time it runs, for further iterations you need to change the code for it to actualy do something in the `cargo benchmark` - I would like to update the readme here so that the newbies don't get confused. There isn't any json files being generated or something similar on my machine, the tests just finish with:

```
test result: ok. 0 passed; 0 failed; 162 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

but I don't notice any new file and basically the list script is just showing stuff available in: 
`https://milli-benchmarks.fra1.digitaloceanspaces.com` 

On the other hand I don't have a possibility to trigger actions or something on the repo (and this is fine since I'm not maintainer).

Do I need to reach out to the maintainers to compare performance of this branch to main or something similar?

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Hey @msvaljek,

Thank you for participating! Unfortunately, the soft and hard maps combo was designed to store "soft-deleted" ids in the soft map. You must remove everything that refers to the `DELETED_ID` variable and make sure that when you delete ids from the, now, single map they are effectively removed by reconstructing the map entirely i.e. what we do in `insert_ids`.

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
Hello @msvaljek 
If you are still around, can you let us know if you plan to finish the PR? Indeed, the tests still fail and we cannot merge or even review it more before it's fixed 😊

**Comment by msvaljek** [comment]
> * Tests on ubuntu-20.04

I'm still working on this, I'll try to carve out some time latest this weekend.

**Comment by msvaljek** [comment]
> Hello @msvaljek If you are still around, can you let us know if you plan to finish the PR? Indeed, the tests still fail and we cannot merge or even review it more before it's fixed 😊

I'm still working on this, I'll try to carve out some time latest this weekend.

File: milli/src/external_documents_ids.rs [hunk 1]
```diff
@@ -49,30 +44,24 @@ impl<'a> ExternalDocumentsIds<'a> {
...
         while let Some((external_id, docids)) = iter.next() {
-            if docids.iter().any(|v| v.index == 1) {
-                // If the `other` set returns a value here it means
-                // that it must be marked as deleted.
-                new_soft_builder.insert(external_id, DELETED_ID)?;
-            } else {
-                let value = docids.iter().find(|v| v.index == 0).unwrap().value;
-                new_soft_builder.insert(external_id, value)?;
-            }
+            let value = docids.iter().find(|v| v.index == 0).unwrap().value;
```
Comment (Kerollmops) [inline milli/src/external_documents_ids.rs:52]: Could you add a comment above this `iter.find` call that explains what you are trying to retrieve, please?

**Review by Kerollmops — CHANGES_REQUESTED** [review]
It looks pretty good to me. Thank you very much for that.

I am just wondering if you could send a dataset in multiple batches multiple times and see the time difference between the previous version and the new one. I am sure the recently introduced soft-deletion feature will amortize it, but who knows?

**Comment by msvaljek** [comment]
> It looks pretty good to me. Thank you very much for that.
> 
> I am just wondering if you could send a dataset in multiple batches multiple times and see the time difference between the previous version and the new one. I am sure the recently introduced soft-deletion feature will amortize it, but who knows?

I [already explored](https://github.com/meilisearch/milli/pull/666#issuecomment-1281444715) and I don't think I have a permission to run this.

**Review by Kerollmops — COMMENTED** [review]
Would it be possible that you set up a Meilisearch [with a maximum index size](https://docs.meilisearch.com/learn/configuration/instance_options.html#max-index-size) of something like 5GiB, [send the movies dataset](https://docs.meilisearch.com/learn/getting_started/quick_start.html#add-documents) in multiple batches (by batches of 100 docs) on both your branch and on [the latest release of v0.30.0](https://github.com/meilisearch/meilisearch/releases), please?

Just to see a possible perf issue.

**Comment by msvaljek** [comment]
The implementations diverged a lot since I had time to work on this one, I'll reopen the PR again when I resolve the differences.

**Pull request closed without merging**